src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs

```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
             return policiesArray;
         }
 
-        private SafeCreateHandle PrepareCertsArray(ICertificatePal cert, X509Certificate2Collection extraStore)
+        private SafeCreateHandle PrepareCertsArray(
+            ICertificatePal cert,
+            X509Certificate2Collection extraStore,
+            X509Certificate2Collection customTrustStore,
+            X509ChainTrustMode trustMode)
         {
-            IntPtr[] ptrs = new IntPtr[1 + (extraStore?.Count ?? 0)];
-            SafeHandle[] safeHandles = new SafeHandle[ptrs.Length];
-
+            int certificateCount = 1 + (extraStore?.Count ?? 0) + (trustMode == X509ChainTrustMode.CustomRootTrust ? customTrustStore?.Count ?? 0 : 0);
+            int certificatesAdded = 1;
+            SafeHandle[] safeHandles = new SafeHandle[certificateCount];
             AppleCertificatePal applePal = (AppleCertificatePal)cert;
-
             safeHandles[0] = applePal.CertificateHandle;
 
             if (extraStore != null)
             {
-                for (int i = 0; i < extraStore.Count; i++)
+                foreach (X509Certificate2 extraCert in extraStore)
                 {
-                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraStore[i].Pal;
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraCert.Pal;
+                    safeHandles[certificatesAdded] = extraCertPal.CertificateHandle;
+                    certificatesAdded++;
+                }
+            }
 
-                    safeHandles[i + 1] = extraCertPal.CertificateHandle;
+            if (trustMode == X509ChainTrustMode.CustomRootTrust && customTrustStore != null)
+            {
+                for (int i = 0; i < customTrustStore.Count; i++)
+                {
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)customTrustStore[i].Pal;
+                    safeHandles[certificatesAdded + i] = extraCertPal.CertificateHandle;
```

Comment:
Instead of `certificatesAdded + i`, how about instead doing `certificatesAdded++`, and then down below before we return, doing `Debug.Assert(certificatesAdded == certificateCount);`?